File: hello-grpc-csharp/Common/LoggingConfig.cs
using System;
using System.IO;
using log4net;
using log4net.Appender;
using log4net.Core;
using log4net.Layout;
using log4net.Repository.Hierarchy;

namespace Common
{
    /// <summary>
    /// Logging configuration for standardized logging setup.
    /// Provides utilities to initialize logging with standard format:
    /// [TIMESTAMP] [LEVEL] [COMPONENT] MESSAGE [CONTEXT]
    /// </summary>
    public static class LoggingConfig
    {
        private const string DefaultLogDir = "logs";
        private const string TimestampFormat = "yyyy-MM-dd HH:mm:ss.fff";
        private const string LogPattern = "[%date{yyyy-MM-dd HH:mm:ss.fff}] [%level] [%logger] %message%newline";

        /// <summary>
        /// Initialize logging for a component with dual output (console and file)
        /// </summary>
        /// <param name="component">The component name (e.g., "client", "server")</param>
        /// <param name="logDir">The directory for log files (default: "logs")</param>
        /// <returns>Logger instance for the component</returns>
        public static ILog InitializeLogging(string component, string logDir = DefaultLogDir)
        {
            // Create log directory
            Directory.CreateDirectory(logDir);

            // Get the repository
            var repository = LogManager.GetRepository() as Hierarchy;
            if (repository == null)
            {
                throw new InvalidOperationException("Failed to get log4net repository");
            }

            // Create pattern layout
            var patternLayout = new PatternLayout
            {
                ConversionPattern = LogPattern
            };
            patternLayout.ActivateOptions();

            // Create console appender
            var consoleAppender = new ConsoleAppender
            {
                Layout = patternLayout,
                Threshold = GetLogLevel()
            };
            consoleAppender.ActivateOptions();

            // Create file appender with timestamp
            var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
            var logFileName = Path.Combine(logDir, $"{component}_{timestamp}.log");
            
            var fileAppender = new RollingFileAppender
            {
                File = logFileName,
                AppendToFile = true,
                RollingStyle = RollingFileAppender.RollingMode.Size,
                MaxSizeRollBackups = 7,
                MaximumFileSize = "100MB",
                StaticLogFileName = true,
                Layout = patternLayout,
                Threshold = GetLogLevel()
            };
            fileAppender.ActivateOptions();

            // Configure repository
            repository.Root.Level = GetLogLevel();
            repository.Root.AddAppender(consoleAppender);
            repository.Root.AddAppender(fileAppender);
            repository.Configured = true;

            // Get logger
            var logger = LogManager.GetLogger(component);
            logger.Info($"Logging initialized for component: {component}");
            logger.Info($"Log file: {logFileName}");

            return logger;
        }

        /// <summary>
        /// Get log level from environment variable or default to INFO
        /// </summary>
        /// <returns>The log level</returns>
        private static Level GetLogLevel()
        {
            var levelStr = Environment.GetEnvironmentVariable("LOG_LEVEL");
            if (string.IsNullOrEmpty(levelStr))
            {
                return Level.Info;
            }

            return levelStr.ToUpper() switch
            {
                "DEBUG" => Level.Debug,
                "INFO" => Level.Info,
                "WARN" => Level.Warn,
                "WARNING" => Level.Warn,
                "ERROR" => Level.Error,
                "FATAL" => Level.Fatal,
                _ => Level.Info
            };
        }

        /// <summary>
        /// Shutdown logging and flush all appenders
        /// </summary>
        public static void ShutdownLogging()
        {
            LogManager.Shutdown();
        }
    }
}

File: hello-grpc-java/src/main/java/org/feuyeux/grpc/common/LoggingConfig.java
package org.feuyeux.grpc.common;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Logging configuration helper for standardized logging setup.
 *
 * <p>Provides utilities to initialize logging with standard format: [TIMESTAMP] [LEVEL] [COMPONENT]
 * MESSAGE [CONTEXT]
 *
 * <p>Logging is configured via logback.xml with dual output (console and file).
 */
public class LoggingConfig {
  private static final String DEFAULT_LOG_DIR = "logs";
  private static final DateTimeFormatter TIMESTAMP_FORMAT =
      DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss");

  /**
   * Initialize logging for a component.
   *
   * <p>Creates log directory if it doesn't exist and sets system properties for logback
   * configuration.
   *
   * @param component The component name (e.g., "client", "server")
   * @return Logger instance for the component
   */
  public static Logger initializeLogging(String component) {
    return initializeLogging(component, DEFAULT_LOG_DIR);
  }

  /**
   * Initialize logging for a component with custom log directory.
   *
   * @param component The component name (e.g., "client", "server")
   * @param logDir The directory for log files
   * @return Logger instance for the component
   */
  public static Logger initializeLogging(String component, String logDir) {
    // Create log directory
    try {
      Path logPath = Paths.get(logDir);
      if (!Files.exists(logPath)) {
        Files.createDirectories(logPath);
      }
    } catch (IOException e) {
      System.err.println("Failed to create log directory: " + e.getMessage());
    }

    // Set system properties for logback
    System.setProperty("COMPONENT", component);
    String timestamp = LocalDateTime.now().format(TIMESTAMP_FORMAT);
    System.setProperty("bySecond", timestamp);

    // Get logger
    Logger logger = LoggerFactory.getLogger(component);
    logger.info("Logging initialized for component: {}", component);

    return logger;
  }

  /**
   * Get log level from environment variable or default to INFO.
   *
   * @return The log level string
   */
  public static String getLogLevel() {
    String level = System.getenv("LOG_LEVEL");
    return level != null ? level : "INFO";
  }
}
